soccer_ipm/test/test_ipm.py

```diff
@@ -0,0 +1,155 @@
+    received_msg: List[Optional[output_msg_type]] = [None]
+
+    # Create a callback with sets this reference
+    def callback(msg):
+        received_msg[0] = msg
+
+    # Subscribe to IPM results
+    test_node.create_subscription(
+        output_msg_type, output_topic, callback, 10)
+
+    # Create header message for the current time stamp in the camera frame
+    header = Header(
+        stamp=node.get_clock().now().to_msg(),
+        frame_id='camera_optical_frame')
+
+    # Create a dummy transform from the camera to the base_footprint frame
+    tf = TransformStamped(
+        header=header,
+        child_frame_id='base_footprint',
+    )
+    tf.transform.translation.z = 1.0
+    tf.transform.rotation.x = 0.0
+    tf.transform.rotation.w = 1.0
+
+    # Publish the dummy transform
+    tf_pub.publish(TFMessage(
+        transforms=[tf]
+    ))
+    # Spin the ipm to process the new data
+    rclpy.spin_once(node, timeout_sec=0.1)
+
+    # Send camera info message to the IPM
+    camera_info.header.stamp = header.stamp
+    camera_info_pub.publish(camera_info)
+    # Spin the IPM to process the new data
+    rclpy.spin_once(node, timeout_sec=0.1)
+
+    # Send image space detection
+    input_msg.header = header
+    ball_pub.publish(input_msg)
+    # Spin the IPM to process the new data
+    rclpy.spin_once(node, timeout_sec=0.1)
+
+    # Spin the test__node to recive the results from the IPM
+    rclpy.spin_once(test_node, timeout_sec=0.1)
+
+    # Assert that we recived a message
+    assert received_msg[0] is not None
+
+    # Clean shutdown of the nodes
+    rclpy.shutdown()
+    node.destroy_node()
+    test_node.destroy_node()
+
+    return received_msg[0], input_msg
+
+
+def test_ipm_empty_ball():
+    standard_ipm_test_case(
+        BallArray,
+        'balls_in_image',
+        BallArray(),
+        sv3dm.BallArray,
+        'balls_relative')
+
+
+def test_ipm_ball():
+    # Create ball detection
+    ball = Ball()
+    ball.center.x = camera_info.width / camera_info.binning_x // 2
+    ball.center.y = camera_info.height / camera_info.binning_y // 2
+    ball.confidence = Confidence(confidence=0.42)
+    ball_detection = BallArray(balls=[ball])
+
+    out, inp = standard_ipm_test_case(
+        BallArray,
+        'balls_in_image',
+        ball_detection,
+        sv3dm.BallArray,
+        'balls_relative')
+
+    # Assert that we recived the correct message
+    assert len(out.balls) == 1, 'Got too many balls'
+    assert out.header.stamp == inp.header.stamp, 'Time stamp got changed by the ipm'
+    assert out.header.frame_id == 'base_footprint', \
+        'Output frame is not "base_footprint"'
+    ball_relative: sv3dm.Ball = out.balls[0]
+    np.testing.assert_allclose(
+        ball_relative.confidence.confidence,
+        ball.confidence.confidence)
+    np.testing.assert_allclose(
+        [ball_relative.center.x, ball_relative.center.y, ball_relative.center.z],
+        [0.0, 0.0, 0.0])
```

Comment:
More code coverage for testing the other msg types seems needed here.

Reply:
Yeah

Reply:
Now all soccer vision types are implemented and have some test coverage. But I will add a few more cases for the markings.